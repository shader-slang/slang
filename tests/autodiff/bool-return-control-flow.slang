//TEST(compute):COMPARE_COMPUTE_EX:-slang -compute -shaderobj -output-using-type
//TEST(compute, vulkan):COMPARE_COMPUTE_EX:-vk -compute -shaderobj -output-using-type

//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name=outputBuffer
RWStructuredBuffer<float> outputBuffer;

[BackwardDifferentiable]
bool conditionFunc(no_diff float a, inout float x)
{
    x = x * a;
    return x > 100.f;
}

[BackwardDifferentiable]
float outerFunc(no_diff float a, float x)
{
    if (conditionFunc(a, x))
        return x;
    else
        return -x;
}

[numthreads(1, 1, 1)]
void computeMain(uint3 dispatchThreadID: SV_DispatchThreadID)
{
    float a = 10.0;
    DifferentialPair<float> dpx = DifferentialPair<float>(4.f, 1.f);
    __bwd_diff(outerFunc)(a, dpx, 1.0);

    outputBuffer[0] = dpx.d;
}