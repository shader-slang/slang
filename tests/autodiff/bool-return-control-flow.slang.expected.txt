type: float
-10.000000
0.000000
0.000000
0.000000